### PR TITLE
Display serverspec Results to a template using JSON

### PR DESCRIPTION
**Pull request by Joeper214**
### Objectives of this PR

- [x] Create templates for Serverspec Result Display.
- [x] Query from Activemodel the needed serverspec result data.
- [x] Create Serverspec results JS and JSON builder file.
- [x] Display/ Map the Serverspec result data to the Template.

### CODE REVIEW

**Comment by Joeper214** [comment]
Test is broken again. 
sad

**Comment by Joeper214** [comment]
test is broken again. :(

**Comment by root-beer** [comment]
Test coverage decreased but test passed.
No problem. :)

**Comment by Joeper214** [comment]
Can you help me fix the test? :)

**Comment by root-beer** [comment]
Sorry. I'm not good at test.

**Comment by Joeper214** [comment]
It's okay.

**Comment by Joeper214** [comment]
Test is broken. Please help. @pocke

**Comment by pocke** [comment]
@Joeper214 What should `logs` action assign to `@serverspec_results` ?

**Comment by Joeper214** [comment]
Logs are the serverspec results retrieval data. from model to displaying of results to the html template.

**Comment by Joeper214** [comment]
Sorry. I guess my naming convention is quite confusing. I'll fix this.

**Comment by pocke** [comment]
Thanks

**Comment by pocke** [comment]
I Fixed spec, but rspec has been broken yet.
Please try to fix?

**Comment by Joeper214** [comment]
Do you know how to create a test for this query? 
```
  @serverspec_results = ServerspecResult.joins(:serverspec_result_details, :serverspecs, :resource).where(resource_id: resource.id)
    .group('serverspec_result_details.id')
    .order('serverspec_results.id DESC')
    .select("serverspec_results.*, serverspecs.name, resources.physical_id, serverspec_result_details.id as 'result_id'")
```

**Comment by pocke** [comment]
I'll Investigate this

**Comment by Joeper214** [comment]
Thanks!

**Comment by pocke** [comment]
Can we replace like the following code?

```ruby
# Before
@serverspec_results = ServerspecResult.joins(:serverspec_result_details, :serverspecs, :resource).where(resource_id: resource.id)
.group('serverspec_result_details.id')
.order('serverspec_results.id DESC')
.select("serverspec_results.*, serverspecs.name, resources.physical_id, serverspec_result_details.id as 'result_id'")

# => [#<ServerspecResult:0x0000000a883830 id: 1, resource_id: 1, status: 0, created_at: Tue, 08 Sep 2015 13:36:05 JST +09:00, updated_at: Tue, 08 Sep 2015 13:36:05 JST +09:00, message: "serverspec success">,
#     #<ServerspecResult:0x0000000a8836f0 id: 1, resource_id: 1, status: 0, created_at: Tue, 08 Sep 2015 13:36:05 JST +09:00, updated_at: Tue, 08 Sep 2015 13:36:05 JST +09:00, message: "serverspec success">,
#     #<ServerspecResult:0x0000000a8834c0 id: 1, resource_id: 1, status: 0, created_at: Tue, 08 Sep 2015 13:36:05 JST +09:00, updated_at: Tue, 08 Sep 2015 13:36:05 JST +09:00, message: "serverspec success">]

# After
@serverspec_results = resource.serverspec_results
# => [#<ServerspecResult:0x00000009c0d6a0 id: 1, resource_id: 1, status: 0, created_at: Tue, 08 Sep 2015 13:36:05 JST +09:00, updated_at: Tue, 08 Sep 2015 13:36:05 JST +09:00, message: "serverspec success">]

```

Result of after code is unique.

**Comment by pocke** [comment]
I replaced SQL query at 7a000bb .

**Comment by Joeper214** [comment]
I think it does not returns the serverspec name and physical_id of the resource.

**Comment by Joeper214** [comment]
also, the where clause for a specific resource_id is gone. for querying the specific serverspec result.

**Comment by Joeper214** [comment]
Which causes to type mismatch for this . https://github.com/skyarch-networks/skyhopper/blob/disp-svc-result/app/views/serverspecs/results.jbuilder

**Comment by pocke** [comment]
sorry, I forgot it.
use `includes()` in controller.
http://apidock.com/rails/ActiveRecord/QueryMethods/includes

and the following jbuilder
```ruby
json.array!(@serverspec_results) do |log|
  json.extract! log, :id, :result_id, :physical_id, :message, :status, :created_at, :name
  json.physical_id log.resource.physical_id
end
```

I think it work.

**Comment by pocke** [comment]
http://railscasts.com/episodes/320-jbuilder?view=asciicast
Jbuilder sample code

```ruby
json.(@article, :id, :name, :published_at)
json.edit_url edit_article_url(@article) if current_user.admin?
```

**Comment by Joeper214** [comment]
Thanks! I will try.

**Comment by Joeper214** [comment]
I'll fix conflicts.

**Comment by Joeper214** [comment]
Please review and PR. Thanks!

File: app/assets/javascripts/modules/listen.js [hunk 1]
```diff
@@ -39,11 +41,18 @@ function render_infrastructures(value, key, lang){
...
     return ret;
+  }else if (key === 'status') {
+    if(value == "CREATE_COMPLETE"){
+      return "<span class='text text-success'>"+value+"</span>";
+    }else if (value === 'DELETE_IN_PROGRESS') {
+      return "<span class='text text-danger'>"+value+"</span>";
+    }else if (value === 'CREATE_IN_PROGRESS') {
+      return "<span class='text text-info'>"+value+"</span>";
+    }else{
+      return 'NO STACK INFO';
```
Comment (Joeper214) [inline app/assets/javascripts/modules/listen.js]: Sorry I think this causes problems. I'll fix this.

**Comment by Joeper214** [comment]
Please Merge PR after reviewing it. Thanks!

**Comment by pocke** [comment]
I start reviewing.

**Comment by Joeper214** [comment]
Thanks!

File: app/assets/javascripts/infrastructures.js [hunk 2]
```diff
@@ -36,6 +36,7 @@
   var listen = require('./modules/listen');
   var parseURLParams = require('./modules/getURL');
   var infraindex = require('./modules/loadindex');
+  var http = require('http');
```
Comment (pocke) [inline app/assets/javascripts/infrastructures.js]: Why do you need?
I can't find code of using `http` module.

Comment (Joeper214) [inline app/assets/javascripts/infrastructures.js]: Sorry I forgot to remove.

File: app/controllers/serverspecs_controller.rb [hunk 3]
```diff
@@ -123,22 +138,20 @@ def run
...
+      ServerspecResult.create(resource_id: resource.id, status: resp[:status_text],message: resp[:message], serverspec_ids: serverspec_ids)
       render_msg = I18n.t('serverspecs.msg.failure', physical_id: physical_id, failure_specs: resp[:message])
     end
 
-      ServerspecResult.create(
-        resource_id: resource.id,
-        status: resp[:status_text],
-        message: resp[:message],
-        serverspec_ids: serverspec_ids
-      )
```
Comment (pocke) [inline app/controllers/serverspecs_controller.rb]: Why did you move this code in case-when expression?
I this previous code is DRY and simpler than current code.

Comment (Joeper214) [inline app/controllers/serverspecs_controller.rb]: :+1:   It seems to be a mistake. I was just trying to fetch the result of the success message. Sorry.

Comment (Joeper214) [inline app/controllers/serverspecs_controller.rb]: I'll return the previous code.

File: app/assets/javascripts/infrastructures.js [hunk 4]
```diff
@@ -1357,16 +1358,94 @@
...
+    filters:{
+      wrap: wrap,
+      listen: listen,
+      paginate: function(list) {
+        var index = this.pageNumber * this.pages;
+        return list.slice(index, index + this.pages);
+      },
+      roundup: function (val) { return (Math.ceil(val));},
+    }
+  });
```
Comment (pocke) [inline app/assets/javascripts/infrastructures.js]: I like the following style about order of properties of Vue.
http://vuejs.org/api/options.html#Lifecycle

```javascript
Vue.component("xxx", {
  template: ...,
  data: ...,
  methods: ...,
  computed: ...,
  filters: ...,
  created: ...,
  compiled: ...,
  ready: ...,
});
```

I want lifecycle methods to be in one place, and ordered by execution sequence.


Could you move `compiled` property to under the created ?

Comment (Joeper214) [inline app/assets/javascripts/infrastructures.js]: okay. No problem. :)

File: app/controllers/serverspecs_controller.rb (hunk 3 above)
Comment (pocke) [inline app/controllers/serverspecs_controller.rb]: Thanks :+1:

**Comment by pocke** [comment]
I want back button at result page.

default


https://github.com/skyarch-networks/skyhopper/blob/6277294e56eb53daed99fd346e40805ba0c332e7/config/locales/en.bootstrap.yml#L8
And you can use the above for I18n.

**Comment by Joeper214** [comment]
Okay. Thanks I forgot. but seems I18n is in english?

**Comment by pocke** [comment]
Sorry for not explaining enough.
It means to use `t('helpers.link.back')`.
Japanese is in here. https://github.com/skyarch-networks/skyhopper/blob/6277294e56eb53daed99fd346e40805ba0c332e7/config/locales/ja.bootstrap.yml#L8

**Comment by Joeper214** [comment]
Okay. Thanks I'll implement this immediately. because this PR is taking too long.

**Comment by pocke** [comment]
Thanks!

**Comment by pocke** [comment]
My TODOs

- [ ] Check cascade

**Comment by pocke** [comment]
I think DB scheme should be changed.
However, that is not required, we'll work for that after merged this PR.

**Comment by pocke** [comment]
default

message is null :crying_cat_face: 
I think if message is null, hide :mag: .

File: app/assets/javascripts/modules/listen.js [hunk 5]
```diff
@@ -183,3 +192,48 @@ function render_user_admin(value, key, lang){
...
+    var ret;
+    switch (value) {
+      case 'success':
+        ret = "<span class='label label-success'>"+value+"</span>";
+        break;
+      case 'failed':
+        ret = "<span class='label label-danger'>"+value+"</span>";
+        break;
+      case 'pending':
+        ret = "<span class='label label-info'>"+value+"</span>";
```
Comment (pocke) [inline app/assets/javascripts/modules/listen.js]: `label-warn` is suitable than `label-info` for colour of pending.

Comment (Joeper214) [inline app/assets/javascripts/modules/listen.js]: :+1:

**Comment by pocke** [comment]
Could you reverse default order of results ? 
Order of Serverspec results is inverse of Operation log.

default

default

**Comment by pocke** [comment]
Back button :+1:

**Comment by Joeper214** [comment]
Please try pull again the latest push because I can't seem to replicated this issue.


wewewewe


So that I could see if the fix is working. thanks!

**Comment by pocke** [comment]
Thanks! merged it!
